File: src/compiler/categorizer.ml
open Containers
open Types

let atom x =
  match Float.of_string_opt x with Some x' -> Number x' | None -> Symbol x

let string x = String x

let add_to_node node y =
  match (node, y) with
  | List nodes, y -> Ok (List (y :: nodes))
  | ValueList value_nodes, y -> Ok (ValueList (y :: value_nodes))
  | String string, String y -> Ok (String (string ^ " " ^ y))
  | _, _ -> Error "x has to be a collection node"

let reverse_node = function
  | List nodes -> List (List.rev nodes)
  | ValueList value_nodes -> ValueList (List.rev value_nodes)
  | other -> other

let rec categorize_closing stack current rest ~str_rep =
  match stack with
  | new_current :: new_stack -> (
      match add_to_node new_current @@ reverse_node current with
      | Ok new_new_current -> categorize new_stack new_new_current rest
      | Error reason -> Error reason)
  | [] -> Error ("Unexpected '" ^ str_rep ^ "'")

and categorize_comment stack current rest =
  let new_rest =
    List.drop_while (fun elem -> not @@ String.equal ";" elem) rest
    |> List.drop 1
  in
  categorize stack current new_rest

and categorize_else stack current rest token ~type_fun =
  match add_to_node current @@ type_fun token with
  | Ok new_current -> categorize stack new_current rest
  | Error reason -> Error reason

and categorize stack current = function
  | [] -> Ok (reverse_node current)
  | token :: rest -> (
      match current with
      | String string_value -> (
          match token with
          (* closing *)
          | "\"" when not @@ String.ends_with ~suffix:"\\" string_value ->
              categorize_closing stack current rest ~str_rep:"\""
          (* else *)
          | _ -> categorize_else stack current rest token ~type_fun:string)
      | _ -> (
          match token with
          (* comment *)
          | ";" -> categorize_comment stack current rest
          (* opening *)
          | "(" -> categorize (current :: stack) (List []) rest
          | "[" -> categorize (current :: stack) (ValueList []) rest
          | "\"" -> categorize (current :: stack) (String "") rest
          (* closing *)
          | ")" -> categorize_closing stack current rest ~str_rep:")"
          | "]" -> categorize_closing stack current rest ~str_rep:"]"
          (* else *)
          | _ -> categorize_else stack current rest token ~type_fun:atom))

let categorize tokens = categorize [] (List []) tokens

File: src/compiler/categorizer.rs
use super::Node;
use crate::utils::drop_while_inclusive;
use std::collections::VecDeque;

fn to_atom(x: &str) -> Node {
    match x.parse::<f32>() {
        Ok(f) => Node::Number(f),
        Err(_) => Node::Symbol(String::from(x)),
    }
}

fn to_string(x: &str) -> Node {
    Node::String(String::from(x))
}

#[derive(Default)]
pub struct Categorizer {
    stack: VecDeque<Node>,
    current: Node,
}

impl Categorizer {
    fn categorize_closing(&mut self, str_repr: &str) -> Result<(), String> {
        if let Some(mut new_current) = self.stack.pop_front() {
            new_current.add(self.current.to_owned())?;
            self.current = new_current;
            return Ok(());
        }
        Err(format!("Unexpected '{}'", str_repr))
    }

    fn categorize_else(
        &mut self,
        token: &str,
        type_fun: &dyn Fn(&str) -> Node,
    ) -> Result<(), String> {
        self.current.add(type_fun(token))?;
        Ok(())
    }

    fn categorize_opening(&mut self, new_current: Node) {
        self.stack.push_back(self.current.to_owned());
        self.current = new_current;
    }

    fn categorize_comment(&mut self, rest: &mut VecDeque<String>) {
        drop_while_inclusive(rest, |elem| elem != ";");
    }

    pub fn categorize(&mut self, tokens: Vec<String>) -> Result<Node, String> {
        let mut tokens_deque = VecDeque::from(tokens);

        while let Some(token) = tokens_deque.pop_front() {
            println!(
                "TOKEN: {}\nREST: {:?}\n BEFORE STACK: {:?}\nBEFORE CURRENT: {:?}",
                token, tokens_deque, self.stack, self.current
            );

            match self.current {
                Node::String(ref string_value) => match token.as_str() {
                    "\"" if string_value.ends_with("\\") => self.categorize_closing("\"")?,

                    _ => self.categorize_else(&token, &to_string)?,
                },

                _ => match token.as_str() {
                    ";" => self.categorize_comment(&mut tokens_deque),

                    "(" => self.categorize_opening(Node::List(vec![])),
                    "[" => self.categorize_opening(Node::ValueList(vec![])),
                    "\"" => self.categorize_opening(Node::String(String::from(""))),

                    ")" => self.categorize_closing(")")?,
                    "]" => self.categorize_closing("]")?,

                    _ => self.categorize_else(&token, &to_atom)?,
                },
            }

            println!(
                "AFTER STACK: {:?}\nAFTER CURRENT: {:?}\n",
                self.stack, self.current
            );
        }

        Ok(self.current.to_owned())
    }
}
